ruma-federation-api/src/transactions/send_transaction_message/edu.rs

```diff
@@ -0,0 +1,294 @@
+
+use std::collections::BTreeMap;
+
+use js_int::UInt;
+use ruma_common::{encryption::DeviceKeys, presence::PresenceState};
+use ruma_events::{from_raw_json_value, receipt::Receipt};
+use ruma_identifiers::{DeviceIdBox, EventId, RoomId, UserId};
+use serde::{de, Deserialize, Serialize};
+use serde_json::{value::RawValue as RawJsonValue, Value as JsonValue};
+
+// There is one more edu_type synapse recognizes with the note:
+// FIXME: switch to "m.signing_key_update" when MSC1756 is merged into the
+// spec from "org.matrix.signing_key_update"
+
+/// Type for passing ephemeral data to homeservers.
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "edu_type", content = "content")]
+pub enum Edu {
+    /// An EDU representing presence updates for users of the sending homeserver.
+    #[serde(rename = "m.presence")]
+    Presence(PresenceContent),
+
+    /// An EDU representing receipt updates for users of the sending homeserver.
+    #[serde(rename = "m.receipt")]
+    Receipt(ReceiptContent),
+
+    /// A typing notification EDU for a user in a room.
+    #[serde(rename = "m.typing")]
+    Typing(TypingContent),
+
+    /// An EDU that lets servers push details to each other when one of their users adds
+    /// a new device to their account, required for E2E encryption to correctly target the
+    /// current set of devices for a given user.
+    #[serde(rename = "m.device_list_update")]
+    DeviceListUpdate(DeviceListUpdateContent),
+
+    /// An EDU that lets servers push send events directly to a specific device on a
+    /// remote server - for instance, to maintain an Olm E2E encrypted message channel
+    /// between a local and remote device.
+    #[serde(rename = "m.direct_to_device")]
+    DirectToDevice(DirectDeviceContent),
+
+    #[doc(hidden)]
+    Custom(JsonValue),
+}
+
+#[derive(Debug, Deserialize)]
+struct EduTpeDeHelper {
+    /// The message type field
+    edu_type: String,
+    content: Box<RawJsonValue>,
+}
+
+impl<'de> de::Deserialize<'de> for Edu {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let EduTpeDeHelper { edu_type, ref content } = from_raw_json_value(&json)?;
+
+        Ok(match edu_type.as_ref() {
+            "m.presence" => Self::Presence(from_raw_json_value(content)?),
+            "m.receipt" => Self::Receipt(from_raw_json_value(content)?),
+            "m.typing" => Self::Typing(from_raw_json_value(content)?),
+            "m.device_list_update" => Self::DeviceListUpdate(from_raw_json_value(content)?),
+            "m.direct_to_device" => Self::DirectToDevice(from_raw_json_value(content)?),
+            _ => Self::Custom(from_raw_json_value(content)?),
+        })
+    }
+}
+
+/// The content for "m.presence" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PresenceContent {
```

Comment:
All these structs should be non-exhaustive (with the unstable feature to turn that off) and have constructor functions.

Reply:
Is this something that every public enum/struct needs? I was going off of [the message types](https://github.com/ruma/ruma/blob/main/ruma-events/src/room/message.rs#L279), so they eventually need to be fixed to?

Reply:
Yes, see #215.